### PR TITLE
Jsieve 103 - Sieve vacation implementation

### PR DESCRIPTION
**Pull request by chibenwa**
This PR :

 - Propose a SIEVE vacation implementation for JSIEVE
 - Integrates it in JAMES.

See https://github.com/apache/james-project/pull/36

### CODE REVIEW

**Comment by mbaechler** [comment]
looks good overall but needs some work about object encapsulation

File: core/src/main/java/org/apache/jsieve/commands/optional/Vacation.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    private ActionVacation retrieveAction(Arguments arguments) throws SieveException {
+        ArgumentParser argumentParser = new ArgumentParser(arguments.getArgumentList());
+        ActionVacation.ActionVacationBuilder actionVacationBuilder = ActionVacation.builder();
+
+        argumentParser.validateSingleTags();
+        argumentParser.validateTagsWithValue(FROM, SUBJECT, HANDLE, MIME, DAYS, ADDRESSES);
+
+        actionVacationBuilder.addresses(argumentParser.getStringListForTag(ADDRESSES, ADDRESSES_EXCEPTION_MESSAGE));
```
Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/commands/optional/Vacation.java]: why do you repeat the builder, can't you chain the calls ?

File: core/src/main/java/org/apache/jsieve/commands/optional/Vacation.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    protected void validateArguments(Arguments arguments, SieveContext context) throws SieveException {
+        retrieveAction(arguments);
+    }
+
+    private ActionVacation retrieveAction(Arguments arguments) throws SieveException {
+        ArgumentParser argumentParser = new ArgumentParser(arguments.getArgumentList());
+        ActionVacation.ActionVacationBuilder actionVacationBuilder = ActionVacation.builder();
```
Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/commands/optional/Vacation.java]: Why do you create the builder here, you need it two lines later ...

File: core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            return this;
+        }
+
+        public ActionVacationBuilder from(String from) {
+            this.from = from;
+            return this;
+        }
+
+        public ActionVacationBuilder addresses(List<String> addresses) {
+            if (addresses != null) {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java]: you should not do this

File: core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+    private final String subject;
+    private final String from;
+    private final List<String> addresses;
+    private final String handle;
+    private final String reason;
+    private final String mime;
+    private final int duration;
+
+    private ActionVacation(String subject, String from, List<String> addresses, String reason, int duration, String handle, String mime) {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java]: you should make sure the list is immutable

File: core/src/main/java/org/apache/jsieve/utils/ArgumentParser.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        if (remainingArguments.size() > 1) {
+            throw new SyntaxException(exceptionMessage);
+        }
+        if (remainingArguments.size() == 0) {
+            return null;
+        }
+        return retrieveSingleStringValue(remainingArguments.get(0), exceptionMessage);
+    }
+
+    public void validateSingleTags(String... validTags) throws SyntaxException {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/utils/ArgumentParser.java]: the method name is not good enough : what does it actually  check ?

File: core/src/main/java/org/apache/jsieve/utils/ArgumentParser.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        validateTagCollectionWithExpectations(singleTags, validTags);
+    }
+
+    public void validateTagsWithValue(String... validTags) throws SyntaxException {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/utils/ArgumentParser.java]: the method name is not good enough : what does it actually  check ?

File: mailet/src/main/java/org/apache/jsieve/mailet/VacationAction.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        while (enumeration.hasMoreElements()) {
+            String headerName = (String) enumeration.nextElement();
+            if (headerName.startsWith("List-")) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public long getDayDifference(Date date1, Date date2) {
```
Comment (mbaechler) [inline mailet/src/main/java/org/apache/jsieve/mailet/VacationAction.java]: should not be there

File: core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java (hunk 4 above)
Comment (chibenwa) [inline core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java]: How do I make this without Guava ?

Comment (mbaechler) [inline core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java]: Arf, true.
There's `Collections.unmodifiableList(List<? extends T>)`

Comment (chibenwa) [inline core/src/main/java/org/apache/jsieve/mail/optional/ActionVacation.java]: Ok, thank you !

**Comment by mbaechler** [comment]
:+1:

**Pull request closed without merging**